### PR TITLE
Add extendhfxf2 into compiler rt

### PR DESCRIPTION
**Pull request by biabbas**
Retry pr #109090 with updated extendhfxf2 test

### CODE REVIEW

**Comment by biabbas** [comment]
ping @lntue

File: compiler-rt/test/builtins/Unit/extendhfxf2_test.c [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#include <math.h> // for isnan, isinf
+#include <stdio.h>
+
+#if __LDBL_MANT_DIG__ >= 64 && defined(COMPILER_RT_HAS_FLOAT16)
+long double __extendhfxf2(_Float16 f);
+
+int test_extendhfxf2(_Float16 a, long double expected) {
+  long double x = __extendhfxf2(a);
+  __uint16_t *b = (void *)&a;
+  int ret = !(x == expected || (isnan(x) && isnan(expected)) ||
```
Comment (lntue) [inline compiler-rt/test/builtins/Unit/extendhfxf2_test.c:14]: `!((isnan(x) && isnan(expected)) || x == expected)` should be enough.

File: compiler-rt/test/builtins/Unit/extendhfxf2_test.c [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+           *b, x, expected);
+  }
+  return ret;
+}
+
+char assumption_1[sizeof(_Float16) * CHAR_BIT == 16] = {0};
+
+#endif
+
+int main() {
```
Comment (lntue) [inline compiler-rt/test/builtins/Unit/extendhfxf2_test.c:28]: will it be clearer if we just put everything in a single `ifdef ... else ... endif` like:
```
#ifdef __LDBL_MANT_DIG__ >= 64 && defined(COMPILER_RT_HAS_FLOAT16)
...
int main() {
...
}
#else
int main() {
  printf("skipped");
  return 0;
}
#endif
```
WDYT?

Comment (biabbas) [reply compiler-rt/test/builtins/Unit/extendhfxf2_test.c:28]: Yeah. I think this would be clearer.

**Review by lntue — APPROVED** [review]

**Comment by mstorsjo** [comment]
The updated testcase here still fails on Windows in MinGW configurations: https://github.com/mstorsjo/llvm-mingw/actions/runs/11527502938/job/32098789138

```
# .---command stderr------------
# | D:\a\llvm-mingw\llvm-mingw\llvm-project\compiler-rt\test\builtins\Unit\extendhfxf2_test.c:14:3: error: unknown type name '__uint16_t'; did you mean '__uint128_t'?
# |    14 |   __uint16_t *b = (void *)&a;
# |       |   ^~~~~~~~~~
# |       |   __uint128_t
# | note: '__uint128_t' declared here
# | D:\a\llvm-mingw\llvm-mingw\llvm-project\compiler-rt\test\builtins\Unit\extendhfxf2_test.c:19:12: warning: format specifies type 'unsigned int' but the argument has type '__uint128_t' (aka 'unsigned __int128') [-Wformat]
# |    17 |     printf("error in test__extendhfxf2(%#.4x) = %.20Lf, "
# |       |                                        ~~~~~
# |    18 |            "expected %.20Lf\n",
# |    19 |            *b, x, expected);
# |       |            ^~
# | D:\a\llvm-mingw\llvm-mingw\llvm-project\compiler-rt\test\builtins\Unit\extendhfxf2_test.c:48:49: error: use of unknown builtin '__builtin_huge_valf64x' [-Wimplicit-function-declaration]
# |    48 |   if (test_extendhfxf2(__builtin_huge_valf16(), __builtin_huge_valf64x()))
# |       |                                                 ^
# | D:\a\llvm-mingw\llvm-mingw\llvm-project\compiler-rt\test\builtins\Unit\extendhfxf2_test.c:58:37: error: use of unknown builtin '__builtin_nanf64x' [-Wimplicit-function-declaration]
# |    58 |                        (long double)__builtin_nanf64x("")))
# |       |                                     ^
# | 1 warning and 3 errors generated.
# `-----------------------------
```

While sorting this out, can you revert in the meantime?

We'd need to figure out which targets have a builtin type like `__uint16_t` and which don't (or can we just use `uint16_t` by including `stdint.h`? And which targets have `__builtin_huge_valf64x` and `__builtin_nanf64x` - as they're not available for mingw targets at least.

**Comment by rorth** [comment]
The patch likewise broke the [Solaris/amd64 bot](https://lab.llvm.org/staging/#/builders/120/builds/2975).

**Comment by mstorsjo** [comment]
> The patch likewise broke the [Solaris/amd64 bot](https://lab.llvm.org/staging/#/builders/120/builds/2975).

Thanks - I pushed a revert for this now.

File: compiler-rt/lib/builtins/CMakeLists.txt [hunk 3]
```diff
@@ -104,6 +104,7 @@ set(GENERIC_SOURCES
   divti3.c
   extendsfdf2.c
   extendhfsf2.c
+  extendhfxf2.c
```
Comment (arichardson) [inline compiler-rt/lib/builtins/CMakeLists.txt:107]: The xf type is only supported for x86 so it should not be in the generic sources. 

Could you please add me as a reviewer when you submit a new version of this patch.

Comment (biabbas) [reply compiler-rt/lib/builtins/CMakeLists.txt:107]: Sure.

**Pull request merged**